apis/python/src/tiledbsoma/sparse_nd_array.py

```diff
@@ -207,54 +136,53 @@ def read(
         if coords is None:
             coords = (slice(None),)
 
-        with self._ensure_open():
-            A = self._tiledb_obj
-            shape = A.shape
-            sr = self._soma_reader(schema=A.schema)
+        arr = self._handle.reader
+        shape = arr.shape
+        sr = self._soma_reader(schema=arr.schema)
 
-            if not isinstance(coords, (list, tuple)):
-                raise TypeError(
-                    f"coords type {type(coords)} unsupported; expected list or tuple"
-                )
-            if len(coords) < 1 or len(coords) > A.schema.domain.ndim:
-                raise ValueError(
-                    f"coords {coords} must have length between 1 and ndim ({A.schema.domain.ndim}); got {len(coords)}"
-                )
+        if not isinstance(coords, (list, tuple)):
+            raise TypeError(
+                f"coords type {type(coords)} unsupported; expected list or tuple"
+            )
+        if len(coords) < 1 or len(coords) > arr.schema.domain.ndim:
+            raise ValueError(
+                f"coords {coords} must have length between 1 and ndim ({arr.schema.domain.ndim}); got {len(coords)}"
+            )
 
-            for i, coord in enumerate(coords):
-                #                # Example: coords = [None, 3, slice(4,5)]
-                #                # coord takes on values None, 3, and slice(4,5) in this loop body.
-                dim_name = A.schema.domain.dim(i).name
-                if coord is None:
-                    pass  # No constraint; select all in this dimension
-                elif isinstance(coord, int):
-                    sr.set_dim_points(dim_name, [coord])
-                elif isinstance(coord, np.ndarray):
-                    if coord.ndim != 1:
+        for i, coord in enumerate(coords):
+            #                # Example: coords = [None, 3, slice(4,5)]
+            #                # coord takes on values None, 3, and slice(4,5) in this loop body.
+            dim_name = arr.schema.domain.dim(i).name
+            if coord is None:
+                pass  # No constraint; select all in this dimension
+            elif isinstance(coord, int):
+                sr.set_dim_points(dim_name, [coord])
+            elif isinstance(coord, np.ndarray):
+                if coord.ndim != 1:
+                    raise ValueError(
+                        f"only 1D numpy arrays may be used to index; got {coord.ndim}"
+                    )
+                sr.set_dim_points(dim_name, coord)
+            elif isinstance(coord, slice):
+                ned = arr.nonempty_domain()  # None iff the array has no data
```

Comment:
This could cause a major bug due to SOMAReader being handled differently than _handle.reader.

Example:

1. Create new array (_handle has open read and write, but there is no SOMAReader handle).  
2. Write to array (don't close or flush).
4. Call _this_ read method with a coordinate slice:
    * It will create a new SOMAReader
    * It will call _handle.reader.nonempty_domain() *which will return None* because _handle.reader was opened before th write
    * It will therefore skip setting the SOMAReader dim_range
    * and will return the entire array, rather than the slice the user requested

Example of this with DataFrame - where it asks for 2 rows in a slice, but gets all10 back:
```

import pyarrow as pa
import tiledbsoma as soma

path = 'foo'
sz = 10

df = soma.DataFrame.create(
    path,
    schema=pa.schema(
        [
            ("soma_joinid", pa.int64()),
            ("label", pa.large_string()),
        ]
    ),
    index_column_names=["soma_joinid"],
)
df.write(
    pa.Table.from_pydict(
        {
            "soma_joinid": [i for i in range(sz)],
            "label": [str(i) for i in range(sz)],
        }
    )
)

# the result of this slice read is unexpected
# (note lack of flush before doing the read)

print(repr(df.read((slice(1,3),)).concat()))
```